admin/class-yoast-notification.php

```diff
@@ -52,6 +60,165 @@ public function get_id() {
+	}
+
+	/**
+	 * Get the User Meta key to check for dismissal of notification
+	 *
+	 * @return string User Meta Option key that registers dismissal.
+	 */
+	public function get_dismissal_key() {
+		if ( empty( $this->options['dismissal_key'] ) ) {
+			return $this->options['id'];
+		}
+
+		return $this->options['dismissal_key'];
+	}
+
+	/**
+	 * Is this Notification persistent
+	 *
+	 * @return bool True if persistent, False if fire and forget.
+	 */
+	public function is_persistent() {
+		$id = $this->get_id();
+
+		return ! empty( $id );
+	}
+
+	/**
+	 * Check if the notification is relevant for the current user
+	 *
+	 * @return bool True if a user needs to see this Notification, False if not.
+	 */
+	public function display_for_current_user() {
+		// If the notification is for the current page only, always show.
+		if ( ! $this->is_persistent() ) {
+			return true;
+		}
+
+		// If we are not on a WPSEO page and this is required.
+		if ( true === $this->options['wpseo_page_only'] && ! WPSEO_Utils::is_yoast_seo_page() ) {
+			return false;
+		}
+
+		// If the current user doesn't match capabilities.
+		if ( ! $this->match_capabilities() ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Does the current user match required capabilities
+	 *
+	 * @return bool
+	 */
+	public function match_capabilities() {
+		if ( is_multisite() ) {
+			return WPSEO_Utils::grant_access();
+		}
+
+		/**
+		 * Filter capabilities that enable the displaying of this notification.
+		 *
+		 * @since 3.2
+		 *
+		 * @param array              $capabilities The capabilities that must be present for this Notification.
+		 * @param string             $id           The ID of the notification.
+		 * @param Yoast_Notification $notification The notification object.
+		 *
+		 * @return array of capabilities or empty for no restrictions.
+		 */
+		$capabilities = apply_filters( 'wpseo_notification_capabilities', $this->options['capabilities'], $this->options['id'], $this );
+
+		if ( is_string( $capabilities ) && ! empty( $capabilities ) ) {
+			$capabilities = array( $capabilities );
+		}
+
+		// Should be an array.
+		if ( ! is_array( $capabilities ) ) {
+			$capabilities = array();
+		}
+
+		/**
+		 * Filter capability check to enable 'all' or 'any' capabilities.
+		 *
+		 * @since 3.2
+		 *
+		 * @param string             $capability_check The type of check that will be used to determine if an capability is present.
+		 * @param string             $id               The ID of the notification.
+		 * @param Yoast_Notification $notification     The notification object.
+		 *
+		 * @return string 'all' or 'any'.
+		 */
+		$capability_check = apply_filters( 'wpseo_notification_capability_check', $this->options['capability_check'], $this->options['id'], $this );
+
+		if ( ! in_array( $capability_check, array( 'all', 'any' ), true ) ) {
+			$capability_check = 'all';
+		}
+
+		if ( ! empty( $capabilities ) ) {
+
+			foreach ( $capabilities as $capability ) {
```

Comment:
I think it would be more readable doing the following:

```php
if ( 'all' === $capability_check ) {
  $match = array_reduce( $capabilities, function( $match, $capability ) {
    return $match && current_user_can( $capability );
  } );
} else if ( 'any' === $capability_check ) {
  $match = array_reduce( $capabilities, function( $match, $capability ) {
    return $match || current_user_can( $capability );
  }
}

return $match;
```

Of course we cannot use anonymous functions, but you can move these to be private methods inside the class.

Another solution would be:

```php
$matched_capabilities = array_filter( $capabilities, function( $capability ) {
  return current_user_can( $capability );
});

if ( 'all' === $capability_check ) {
  return count( $matched_capabilities ) === count( $capabilities );
} else if ( 'any' === $capability_check ) {
  return 0 < count( $matched_capabilities );
}
```

Or switch statements:

```php
switch ( $capability_check ) {
  case self::CHECK_ANY_CAPS:
    return $this->match_any_caps( $capabilities );
    break;

  default:
  case self::CHECK_ALL_CAPS:
    return $this->match_all_caps( $capabilities );
    break;
}

Or any combination of the above.